filebeat/spooler/spooler.go

```diff
@@ -61,24 +59,32 @@ func (s *Spooler) Start() {
 func (s *Spooler) run() {
 	defer s.wg.Done()
 
-	ticker := time.NewTicker(s.config.idleTimeout / 2)
-
 	logp.Info("Starting spooler: spool_size: %v; idle_timeout: %s",
 		s.config.spoolSize, s.config.idleTimeout)
 
+	timer := time.NewTimer(s.config.idleTimeout)
+
 loop:
 	for {
 		select {
 		case <-s.exit:
-			ticker.Stop()
 			break loop
 		case event := <-s.Channel:
 			if event != nil {
 				s.queue(event)
 			}
-		case <-ticker.C:
-			s.timedFlush()
+
+			// Stop timer and drain channel. See https://golang.org/pkg/time/#Timer.Reset
+			if !timer.Stop() {
+				<-timer.C
+			}
+
+		case <-timer.C:
+			debugf("Flushing spooler because of timeout. Events flushed: %v", len(s.spool))
+			s.flush()
 		}
+
+		timer.Reset(s.config.idleTimeout)
```

Comment:
not sure, but we need to reset always? Even after having received a time event?

In case flush gets blocked, it indeed makes sense to reset timer. but then the timer should be stopped at this place. Otherwise a timer signal is still in queue if flush took very very looong.

Timer should be stopped + reset here only if buffer has been flushed.